### PR TITLE
Feature/fcla 87 consulting prototype 1

### PR DESCRIPTION
**Pull request by sbraggion**

### CODE REVIEW

File: src/main/kotlin/it/unibo/lss/fcla/consulting/common/EventStore.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+    /**
+     * Retrieve the list of all [DomainEvent] of the specified [aggregateId]
+     */
+    override fun getEventsForAggregate(aggregateId: AggregateId): List<DomainEvent> {
+        return events.getOrDefault(aggregateId, listOf())
+    }
+
+    /**
+     * Retrieve the list of all [DomainEvent] of all the [AggregateId]
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/common/EventStore.kt]: Return type is a map, you wrote list

File: src/main/kotlin/it/unibo/lss/fcla/consulting/common/EventSourcedRepository.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    override fun save(aggregate: A) {
+        val uncommittedEvents = aggregate.getUncommittedEvents()
+        uncommittedEvents.forEach { store.saveEvent(aggregateId = aggregate.aggregateId, it) }
+
+        aggregate.clearUncommittedEvents()
+    }
+
+    /**
+     * Retrieve the complete list of all the events for all the aggregates
```
Comment (NarcAle) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/common/EventSourcedRepository.kt]: Return type is a map but you wrote list

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Consulting.kt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * Select which [event] to apply
+     */
+    override fun applyEvent(event: DomainEvent) {
+        when (event) {
+            is ConsultingCreatedEvent -> apply(event)
+            is ConsultingSummaryUpdatedDescriptionEvent -> apply(event)
+            else -> throw IllegalArgumentException() // TODO fixme
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Consulting.kt]: "// TODO fixme" should be removed

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * Value object representing a date in the consulting context
+ */
+data class Date(val year: Int, val month: Int, val day: Int) {
+
+    companion object {
+        const val MIN_YEAR: Int = 1900
+        const val MIN_MONTH: Int = 1
+        const val MAX_MONTH: Int = 12
+        const val MIN_DAY: Int = 1
+        const val MAX_DAY: Int = 31
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt:15]: February has 28 days. You should have used a library to handle Date. Maybe fix this in a future release

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    init {
+        require(year > MIN_YEAR)
+        require(month in MIN_MONTH..MAX_MONTH)
+        require(day in MIN_DAY..MAX_DAY)
+    }
+
+    /**
+     * Check if the current date is equal to [other]
+     */
+    override fun equals(other: Any?): Boolean {
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt:30]: If you override equals you should also override hashcode (minor detail)

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/events/ConsultingCreatedEvent.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import it.unibo.lss.fcla.consulting.domain.consulting.MemberId
+import it.unibo.lss.fcla.consulting.domain.contracts.DomainEvent
+import it.unibo.lss.fcla.consulting.domain.freelancer.FreelancerId
+
+/**
+ * @author Stefano Braggion
+ *
+ * Event representing a created consulting
+ */
+data class ConsultingCreatedEvent(
```
Comment (NarcAle) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/events/ConsultingCreatedEvent.kt:15]: Maybe ConsultingCreated (without the Event suffix) is better? The same apply for the other classes into the event package

File: src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/ConsultingUseCases.kt [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+import it.unibo.lss.fcla.consulting.domain.consulting.createBiomechanicalConsulting
+import it.unibo.lss.fcla.consulting.domain.consulting.createNutritionistConsulting
+import it.unibo.lss.fcla.consulting.domain.consulting.createPhysiotherapyConsulting
+import it.unibo.lss.fcla.consulting.domain.contracts.DomainEvent
+import it.unibo.lss.fcla.consulting.domain.freelancer.FreelancerId
+
+/**
+ * @author Stefano Braggion
+ *
+ *
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/ConsultingUseCases.kt:21]: Missing javadoc

File: src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/FreelancerUseCases.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+import it.unibo.lss.fcla.consulting.domain.consulting.Date
+import it.unibo.lss.fcla.consulting.domain.freelancer.AvailabilityHours
+import it.unibo.lss.fcla.consulting.domain.freelancer.Freelancer
+import it.unibo.lss.fcla.consulting.domain.freelancer.FreelancerId
+import it.unibo.lss.fcla.consulting.domain.freelancer.createAthleticTrainer
+import it.unibo.lss.fcla.consulting.domain.freelancer.createBiomechanical
+import it.unibo.lss.fcla.consulting.domain.freelancer.createNutritionist
+import it.unibo.lss.fcla.consulting.domain.freelancer.createPhysiotherapist
+import java.time.LocalTime
+
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/FreelancerUseCases.kt:13]: Missing javadoc

File: src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/UseCasesExceptions.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
+package it.unibo.lss.fcla.consulting.usecases
+
+import it.unibo.lss.fcla.consulting.domain.exceptions.ConsultingException
+
+/**
+ *
```
Comment (Giordo94) [inline src/main/kotlin/it/unibo/lss/fcla/consulting/usecases/UseCasesExceptions.kt]: Missing javadoc

**Review by Giordo94 — COMMENTED** [review]
Overall there are no big issues

**Review by NarcAle — COMMENTED** [review]
some details to fix

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt (hunk 4 above)
Comment (sbraggion) [reply src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/Date.kt:15]: yes initially i used this custom type, but in the next release it is planned to be changed

File: src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/events/ConsultingCreatedEvent.kt (hunk 6 above)
Comment (sbraggion) [reply src/main/kotlin/it/unibo/lss/fcla/consulting/domain/consulting/events/ConsultingCreatedEvent.kt:15]: it's intentional to have the suffix Event , to distinguish events from commands (probably introduced in the next release)

**Review by Giordo94 — APPROVED** [review]
Approved

**Pull request merged**